### PR TITLE
Improve the vector flow for SPIR-V codegen

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Run vector unroll right after vectorization, that allows unrolling
transfer op before they get lowered. This way the broadcast and
transpose ops generated by lowering are already in the right vetor size.
Start using mlir core onversion for contract ops and remove our ad hoc
lowering.

### CODE REVIEW

File: iree/compiler/Conversion/Common/VectorTransferOptimization.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/VectorTransferOptimization.cpp:66]: Not sure exactly what this pattern is doing. I dont see any unit tests for this either.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I think some lit tests are needed here to make sure this behavior is presistent. I am going to be changing this in a month or so. Would be good to have tests to capture what is expected here.

**Comment by ThomasRaoux** [comment]
> I think some lit tests are needed here to make sure this behavior is presistent. I am going to be changing this in a month or so. Would be good to have tests to capture what is expected here.

You meant for the pattern below or in general? There are quite a few unit test already for the vectorization and vector lowering.

File: iree/compiler/Conversion/Common/VectorTransferOptimization.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/Common/VectorTransferOptimization.cpp:66]: I added a unit test.

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/Common/VectorTransferOptimization.cpp:66]: That is a very specific pattern. I see why you need it though. Shouldnt this just live in core.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:900]: The comment about lit tests was for this case. Making sure that when we vectorize with transposes involved, the correct dimension is vectorized...

**Review by MaheshRavishankar — APPROVED** [review]
Clarified my comment on lit tests. Overall looks fine apart from that.
Stamping this to unblock, but please add the lit tests as mentioned in the comment.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:900]: This is exercised by the matmul vectorization and we have lit test for that. With the new vectorization pushed by Nicolas the Matmul now has the transpose on the transfer read at the beginning and we unroll it like that and lower the transpose in a second step. Removing this code would break the matmul vectorization lit tests for instance. 
Does that make sense? Or do you think we should test it more explicitly?

Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:900]: >That is a very specific pattern. I see why you need it though. Shouldnt this just live in core.

Yes I agree it should live in core as mentioned in the TODO. I expect a bit of discussion on the semantic of ShapeCastOp and so I would rather not be blocked by that at this point but I'm planning to discuss it with Nicolas and others Vector dialect owners once he is back.

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:900]: > This is exercised by the matmul vectorization and we have lit test for that. With the new vectorization pushed by Nicolas the Matmul now has the transpose on the transfer read at the beginning and we unroll it like that and lower the transpose in a second step. Removing this code would break the matmul vectorization lit tests for instance.
> Does that make sense? Or do you think we should test it more explicitly?

Awesome. Thats fine.

File: iree/compiler/Conversion/Common/VectorTransferOptimization.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/Common/VectorTransferOptimization.cpp:75]: `llvm::count_if`?

**Review by antiagainst — APPROVED** [review]

File: iree/compiler/Conversion/Common/VectorTransferOptimization.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/Common/VectorTransferOptimization.cpp]: done.

**Pull request merged**
